Multi-Arch Support

dave-tucker opened the pull request:
Allow libnetwork to be cross compiled on many architectures.
Add dnet binary to the `bin` directory

mrjana commented:
@dave-tucker Is this still experimental or can we start reviewing?

Also needs a rebase.

dave-tucker commented:
Rebasing and testing to see if it's usable yet. Will push shortly

mrjana commented on netutils/utils_linux.go:
```diff
@@ -0,0 +1,29 @@
+// +build linux
```
Now that windows is going to be supported for the daemon and therefore for libnetwork as well, we need a utils_unsupported to atleast to satisfy a build level windows CI task

dave-tucker replied:
Oddly enough, it builds for me without! I can add one if required though

mrjana commented on Makefile:
```diff
@@ -22,7 +23,19 @@ build: ${build_image}.created
+	@mkdir -p "bin"
+	$(shell which godep) go build -o "bin/dnet-linux-amd64" -tags experimental ./cmd/dnet
+
+cross:
```
For my understanding, this target is not completely validated. Correct?

dave-tucker replied:
Ah yes that's a point. I'll do some bash wizardry to fix it.

mrjana commented:
Other than the addition of an unsupported version of the api the diff LGTM

dave-tucker commented:
Rebased and ready for review @mrjana. Note that I've now added this to CircleCI so we should catch any issue that break compilation on other platforms. We should still look at expanding CI to cover Windows and FreeBSD though. Either Jenkins and/or [AppVeyor](http://www.appveyor.com/) for Windows

mrjana commented:
@dave-tucker The builds are failing on this still. Can you please take a look and rebase if necessary?

dave-tucker commented:
@mrjana yep will take a look today

dave-tucker commented:
@mrjana ok now builds are passing :) Let me know if you want me to squash commits before merge.

Now we're now building `dnet` for many archiectures, hopefully we might catch any changes that break compilation

```
linux/amd64...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
linux/386...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
linux/arm...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
windows/amd64...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
windows/386...
/go/bin/godep go build -o "bin/dnet-$GOOS-$GOARCH" -tags libnetwork_discovery ./cmd/dnet
```

mrjana commented:
LGTM

dave-tucker commented:
@mavenugo can I get another LGTM?

aboch commented:
@dave-tucker Can you please rebase

mavenugo commented:
ping @dave-tucker

mrjana commented:
ping @dave-tucker Can you please rebase? This time I promise you we will merge it :-)

dave-tucker commented:
@mrjana @mavenugo 
Now rebased (finally)
Scope has been reduced to just adding the `make cross` target for now. 
Once #770 has merged, we'll be one step closer to adding this to CI

dave-tucker commented on test/integration/dnet/helpers.bash:
```diff
@@ -168,7 +168,7 @@ EOF
 	   -v /usr/local/bin/runc:/usr/local/bin/runc \
 	   -w /go/src/github.com/docker/libnetwork \
-	   mrjana/golang ./cmd/dnet/dnet -d -D ${hopt} -c ${tomlfile}
+	   mrjana/golang ./bin/dnet -d -D ${hopt} -c ${tomlfile}
```
I meant to ask, what's in `mrjana/golang`?
Can we replace with the `libnetwork-build` image?

aboch commented on Makefile:
```diff
@@ -1,42 +1,46 @@
+.PHONY: all all-local build build-local clean cross cross-local check check-code check-format run-tests integration-tests check-local coveralls circle-ci-cross circle-ci-build circle-ci-check circle-ci
 SHELL=/bin/bash
-build_image=libnetworkbuild
+build_image=libnetwork-build
```
I am not sure why #667 removed the dash from the name.
But you may want to check with @LK4D4 before re-adding the `-`.

aboch commented on Makefile:
```diff
@@ -1,42 +1,46 @@
-	@$(shell which godep) go build  ./...
-	@$(shell which godep) go build -o ./cmd/dnet/dnet ./cmd/dnet
+	@mkdir -p "bin"
+	$(shell which godep) go build -o "bin/dnet" ./cmd/dnet
```
+1 on moving dnet to `bin/`

aboch commented:
LGTM
